File: src/main/java/org/freemind/tree/TreeNode.java
package org.freemind.tree;

import java.util.List;

public interface TreeNode<T> {
  TreeNode<T> getChildAt(int index);

  int getChildCount();

  TreeNode<T> getParent();

  int getIndex(TreeNode<T> child);

  boolean getAllowsChildren();

  boolean isLeaf();

  List<TreeNode<T>> children();

  T getValue();
}

File: src/main/kotlin/org/freemind/tree/TreeNode.kt
package org.freemind.tree

interface TreeNode<T> {
  val value: T

  val parent: TreeNode<T>?

  val childCount: Int

  val allowsChildren: Boolean

  val isLeaf: Boolean
    get() = this.childCount == 0

  operator fun get(index: Int): TreeNode<T>?

  operator fun get(child: TreeNode<T>): Int

  val children: java.util.List<out TreeNode<T>?>
}
